File: support_tools/python/upd_commander_checker/src/upd_commander_checker/rule_selection.py
from collections.abc import Iterable

from .finding import Finding


SUPPORTED_RULES = (
    "UPD001",
    "UPD002",
    "UPD101",
    "UPD102",
    "UPD103",
    "UPD201",
    "UPD202",
    "UPD203",
    "UPD301",
    "UPD302",
    "UPD303",
    "UPD401",
    "UPD402",
    "UPD403",
    "UPD404",
    "UPD405",
)
_SUPPORTED_RULE_SET = frozenset(SUPPORTED_RULES)


def normalize_enabled_rules(values: Iterable[str]) -> tuple[str, ...]:
    normalized = tuple(dict.fromkeys(value.strip().upper() for value in values))
    if any(value not in _SUPPORTED_RULE_SET for value in normalized):
        raise ValueError("enabled_rules contains an unsupported UPD code")
    return normalized


def filter_enabled_findings(
    findings: Iterable[Finding], enabled_rules: tuple[str, ...] | None
) -> list[Finding]:
    if enabled_rules is None:
        return list(findings)
    enabled = frozenset(enabled_rules)
    return [finding for finding in findings if finding.code in enabled]

File: support_tools/go/upd_commander_checker/internal/checker/rule_selection.go
package checker

import "fmt"

var SupportedRules = []string{
	"UPD001", "UPD002",
	"UPD101", "UPD102", "UPD103",
	"UPD201", "UPD202", "UPD203",
	"UPD301", "UPD302", "UPD303",
	"UPD401", "UPD402", "UPD403", "UPD404", "UPD405", "UPD406",
}

var supportedRuleSet = func() map[string]struct{} {
	values := make(map[string]struct{}, len(SupportedRules))
	for _, code := range SupportedRules {
		values[code] = struct{}{}
	}
	return values
}()

func ValidateEnabledRules(values []string) error {
	for _, code := range values {
		if _, ok := supportedRuleSet[code]; !ok {
			return fmt.Errorf("unsupported UPD code: %s", code)
		}
	}
	return nil
}

func FilterEnabledFindings(findings []Finding, enabledRules *[]string) []Finding {
	if enabledRules == nil {
		return findings
	}
	enabled := make(map[string]struct{}, len(*enabledRules))
	for _, code := range *enabledRules {
		enabled[code] = struct{}{}
	}
	filtered := make([]Finding, 0, len(findings))
	for _, finding := range findings {
		if _, ok := enabled[finding.Code]; ok {
			filtered = append(filtered, finding)
		}
	}
	return filtered
}
